cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/InMemorySourceControlOperationRunner.java

```diff
@@ -0,0 +1,134 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.sourcecontrol.operationrunner;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.inject.Inject;
+import io.cdap.cdap.proto.ApplicationDetail;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.proto.sourcecontrol.RepositoryConfig;
+import io.cdap.cdap.sourcecontrol.CommitMeta;
+import io.cdap.cdap.sourcecontrol.NoChangesToPushException;
+import io.cdap.cdap.sourcecontrol.RepositoryManager;
+import io.cdap.cdap.sourcecontrol.RepositoryManagerFactory;
+import io.cdap.cdap.sourcecontrol.UnexpectedRepositoryChangesException;
+import org.eclipse.jgit.api.errors.GitAPIException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collections;
+
+/**
+ * In-Memory implementation for {@link SourceControlOperationRunner}.
+ * Runs all git operation inside calling service.
+ */
+public class InMemorySourceControlOperationRunner implements SourceControlOperationRunner {
+  private final RepositoryManagerFactory repoManagerFactory;
+  private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
+  private static final Logger LOG = LoggerFactory.getLogger(InMemorySourceControlOperationRunner.class);
+
+  @Inject
+  public InMemorySourceControlOperationRunner(RepositoryManagerFactory repoManagerFactory) {
+    this.repoManagerFactory = repoManagerFactory;
+  }
+
+  @Override
+  public PushAppResponse push(NamespaceId namespace, RepositoryConfig repoConfig,
+                              ApplicationDetail appToPush, CommitMeta commitDetails)
+    throws Exception {
+    try (RepositoryManager repositoryManager = repoManagerFactory.create(namespace, repoConfig)) {
+      try {
+        repositoryManager.cloneRemote();
+      } catch (Exception e) {
+        throw new PushFailureException("Failed to clone remote repository", e);
+      }
+      LOG.info("Pushing application configs for : {}", appToPush);
+
+      //TODO CDAP-20371: Add retry logic here in case the head at remote moved while we are doing push
+      String pushedCommit = writeAppDetailAndPush(repositoryManager, appToPush, commitDetails);
+
+      try {
+        Path relativeFilePath = getAndValidateAppConfigRelativePath(repositoryManager, appToPush.getName());
+        String gitFileHash = repositoryManager.getFileHash(relativeFilePath, pushedCommit);
+        return new PushAppResponse(appToPush.getName(), appToPush.getAppVersion(), gitFileHash);
+      } catch (Exception e) {
```

Comment:
similar comment, be very thoughtful about exception handling, most of the time we should not be catching and rethrowing.